### PR TITLE
Make IPython custom shell compatible with IPython >= 7.1

### LINKED ISSUE #26924: Make IPython custom shell compatible with IPython >= 7.1
### Actual behavior:
Any command in the IPython console of MantidPlot results in the following error:
```
TypeError                                 Traceback (most recent call last)
TypeError: our_run_code() got an unexpected keyword argument 'async_'
```
Similarly, any command in the IPython console of the Workbench results in the following error:
```
TypeError                                 Traceback (most recent call last)
TypeError: async_run_code() got an unexpected keyword argument 'async_'
```

### Cause:
[IPython.core.interactiveshell.run_code()](https://ipython.readthedocs.io/en/7.1.0/api/generated/IPython.core.interactiveshell.html?highlight=run_code#IPython.core.interactiveshell.InteractiveShell.run_code) for IPython >= 7.1 is not a function by an `asyncio` coroutine

Other useful links:
- [our_run_code()](https://github.com/mantidproject/mantid/blob/master/MantidPlot/ipython_widget/mantid_ipython_widget.py#L39)
-  [async_run_code()](https://github.com/mantidproject/mantid/blob/master/qt/python/mantidqt/widgets/jupyterconsole.py#L74).

### PR DESCRIPTION
**Pull request by jmborr**
**Description of work.**
- [x] method returns a coroutine expected by the ipython shell object
- [x] updated [release notes](https://github.com/mantidproject/mantid/blob/06f2fe59d7a7db81e7a3f4f098a9b034e0b64084/docs/source/release/v4.2.0/framework.rst#python)

**To test:**
test the ipython window for MantidPlot and the Workbench with Python 3.x and IPython >= 7.1. No errors or warnings should pop up.

Fixes #26924

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: MantidPlot/ipython_widget/mantid_ipython_widget.py [hunk 1]
```diff
@@ -47,18 +47,20 @@ def our_run_code(self, code_obj, result=None):
...
     """
-
-    t = threading.Thread()
-    #ipython 3.0 introduces a third argument named result
-    nargs = len(inspect.getargspec(self.ipython_run_code).args)
-    if (nargs == 3):
-        t = threading.Thread(target=self.ipython_run_code, args=[code_obj,result])
+    # Different target arguments depending on IPython's version
+    if 'result' in getfullargspec(self.ipython_run_code).args:
+        if 'async_' in getfullargspec(self.ipython_run_code).kwonlyargs:
```
Comment (martyngigg) [inline MantidPlot/ipython_widget/mantid_ipython_widget.py]: It looks like this is not compatible with Python 2. I get the following error on trying to execute a line of code:
```python
AttributeError                            Traceback (most recent call last)
/home/mgigg/Code/mantidproject/mantid-upstream/qt/python/mantidqt/widgets/jupyterconsole.py in async_run_code(self, code_obj, result, async_)
     88         # Different target arguments depending on IPython's version
     89         if 'result' in getfullargspec(orig_run_code).args:
---> 90             if 'async_' in getfullargspec(orig_run_code).kwonlyargs:
     91                 return orig_run_code(code_obj, result, async_=async_)  # return coroutine to be awaited
     92             else:

AttributeError: 'ArgSpec' object has no attribute 'kwonlyargs'
```

**Review by martyngigg — CHANGES_REQUESTED** [review]
I ran the code inside a virtualenv with IPython 7.8/Python 3 and the code executes correctly.
There is a known warning with `qtconsole` regarding unexpected arguments: jupyter/qtconsole#313 but that can be solved by upgrading `qtconsole`. There's just the issue of Python 2 compatability.

File: MantidPlot/ipython_widget/mantid_ipython_widget.py (hunk 1 above)
Comment (jmborr) [reply MantidPlot/ipython_widget/mantid_ipython_widget.py]: I forgot python 2.x doesn't have keyword-only arguments

Comment (jmborr) [reply MantidPlot/ipython_widget/mantid_ipython_widget.py]: I am on a conference/vacation until next Monday. I doubt I will have enough time to work on this until then

**Comment by jmborr** [comment]
@martyngigg I introduced a small change to deal with python 2.x  and I recompiled the branch against python 2.7. It raised no exceptions .

**Review by martyngigg — APPROVED** [review]
This works on both Python 2 & 3 with the current versions of IPython and IPython 7 as long as `qtconsole` has been upgraded.

**Comment by jmborr** [comment]
@martyngigg @DanNixon Thanks, guys! :smile:

**Pull request merged**